### ISSUE TITLE
Maintaining a proper aspect ratio even when adding a constant frame

### ISSUE THREAD

**Issue by vonas** [darktable-org/darktable#13871]
**Is your feature request related to a problem? Please describe.**
Yes. I am posting photos on Instagram and to limit the height of posts the platform enforces a vertical aspect ratio of 4:5 (WxH). I am also adding white frames to my images with the [framing module](https://docs.darktable.org/usermanual/3.6/en/module-reference/processing-modules/framing/) and I'm cropping images to 4:5 with the [crop module](https://docs.darktable.org/usermanual/3.8/en/module-reference/processing-modules/crop/). The problem here lies in choosing constant borders (the same amount of border thickness on each edge) and wishing to maintain an exported aspect ratio of 4:5.

When I use a constant frame, the aspect ratio of the exported image is inherently changed because a fixed amount of pixels is added to the width and height. With e.g. cropping my photo to 4:5 and then adding a constant border I suddenly get an image that has less height than I could have. If 4:5 is represented by a height multiplier of 1.25 (5 divided by 4) then in the case of a 2451x3063 photo the new aspect ratio is described by the multiplier 1.237... (less height).

Full example: Adding a 4% constant border to an image of dimensions 2451x3063 (4:5, WxH) results in an image of dimensions 2578x3190 (less than 4:5).

**Describe the solution you'd like**
I would like it if the crop module had the option to respect the output of the framing module. That means, I tick a box in the cropping module and the image is cropped in such a way that, including the dimensional changes the framing module makes, the final aspect ratio of the entire image (including frame/s) is the one I have selected in the cropping module. Currently, only the original photo is cropped by that aspect ratio and the frame is added around it, as described above.

**Alternatives**
No alternative suggestions.

**Additional context**
The only solution seems to be to calculate the appropriate height/width that, combined with the desired constant border, results in a 4:5 aspect ratio, and then input that into the crop module (by hand), which is very tedious.

I have calculated the multipliers for height and width when adding a 4% constant border:
* h = 1.0105263157894737 (multiply the height by this)
* w = 0.9895833333333333 (multiply the width by this)

Use these numbers to get the crop dimensions needed to still get a 4:5 aspect ratio when adding a 4% constant frame afterwards (only use one multiplier). This is a workaround and a manual solution.

You can see how many pixels you get from this in this screenshot:

![image](https://user-images.githubusercontent.com/24623262/224105355-edc793f7-a74d-4d24-8a19-8133e80196d9.png)

It's not much but it does add to the overall appearance of the photo.

I would be happy to help implement this, but I have never looked at the code base of Darktable before. Some pointers would be greatly appreciated!

Attached are my notes on calculating the multipliers above.

```
09-03-23

Example:
4000x5000 (HxB, WxH)
add a constant 4% border:
in Darktable that means 4% white space across the height (not width)
new height: 5000 / 0.96 = 5208
5000 * (1/0.96 - 1) = 5000 * 0.041666 = 208.333
new width: 4000 + 208 = 4208
=> 4208x5208
new aspect ratio: 5208/4208 = 1.2376425855513309 (instead of near 1.25)

Make 5000 variable (= h)
since we want to make the resulting image taller (from 1.23 to 1.25),
we need to make the original image taller.
that means h must grow:

(h * 1/0.96) / (4000 + h * (1/0.96 - 1)) = 1.25
(h/0.96)/(4000+h*(1/0.96-1))=1.25
(h/0.96)/(4+h*(1/0.96-1))=1.25

(make 4000 a 4, so we have a multiplier for our height)

(h*1/0.96)/(4+h*(1/0.96-1)) = 1.25  | * (4+h*(1/0.96-1))
h*1/0.96 = 1.25*(4+h*(1/0.96-1))
1/0.96*h = 1.25*4 + 1.25*(1/0.96-1)*h  | - 1.25*(1/0.96-1)*h
h*(1/0.96 - 1.25*(1/0.96-1)) = 1.25*4  | / (1/0.96 - 1.25*(1/0.96-1))
h = 1.25*4 / (1/0.96 - 1.25*(1/0.96-1))

=> h = approx. 5.052631578947369

new aspect ratio: 5.052631578947369:4
= 1.2631578947368423
(or inverted, much nicer = 0.7916666666666665)

now we need to calculate w (= 4) down,
for when we cannot increase the height of the image
but we can only decrease the width:
5 / h = 0.9895833333333333
5/h * 4 = 3.958333333333333

=> h = 5.052631578947369 (increase height)
=> w = 3.958333333333333 (increase width)
(only for 4% frames!)

=> h1 = h/5 = 1.0105263157894737
=> w1 = w/4 = 0.9895833333333333
```

**Comment by parafin** [comment]
Note that your proposed solution only works if you set framing first, crop afterwards and don’t touch framing again.

IMHO your workaround is the solution, I think there is a way to add custom aspect ratio to crop module, so you don’t have to enter it manually each time.

**Comment by MStraeten** [comment]
in your ~/.config/darktable add something like `plugins/darkroom/clipping/extra_aspect_ratios/3.5x4.5=7:9`

**Comment by vonas** [comment]
@parafin @MStraeten Thank you very much! I tried it and it worked like a charm. ([corresponding manual entry](https://docs.darktable.org/usermanual/3.8/en/module-reference/processing-modules/crop/#crop-settings))

In case anyone comes across this issue in the future, this is what I did for reference, to get an aspect ratio that accounts for a 4% constant frame, in order for the exported image to have an exact 4:5 (1,25) aspect ratio:

I added the following line to my `darktablerc` file (`$HOME/.config/darktable/darktablerc`)

```
plugins/darkroom/clipping/extra_aspect_ratios/5:4, 4x5, accounting for 4% frame=505263157:400000000
```

Since it seems an aspect ratio can only consist of integers and not decimal numbers, I multiplied one number from above (5.052631578947369) and 4 by a multiple of 10, so I got integers, but minding that 32-bit signed integers only go up until 2.147.483.647, so there's no overflow.

![grafik](https://user-images.githubusercontent.com/24623262/224488336-734f386d-5136-4260-a6ad-cc8e2361874a.png)

Thanks for the tip! :)
